File: swift/BCCrypto/Tests/BCCryptoTests/ScryptTests.swift
import Foundation
import XCTest
@testable import BCCrypto

final class ScryptTests: XCTestCase {
    func testScryptBasic() {
        let pass = Data("password".utf8)
        let salt = Data("salt".utf8)
        let output = scrypt(password: pass, salt: salt, outputLength: 32)
        XCTAssertEqual(output.count, 32)
        XCTAssertEqual(output, scrypt(password: pass, salt: salt, outputLength: 32))
    }

    func testScryptDifferentSalt() {
        let pass = Data("password".utf8)
        let salt1 = Data("salt1".utf8)
        let salt2 = Data("salt2".utf8)

        XCTAssertNotEqual(
            scrypt(password: pass, salt: salt1, outputLength: 32),
            scrypt(password: pass, salt: salt2, outputLength: 32)
        )
    }

    func testScryptCustomParams() {
        let pass = Data("password".utf8)
        let salt = Data("salt".utf8)
        let output = scrypt(
            password: pass,
            salt: salt,
            outputLength: 32,
            logN: 15,
            r: 8,
            p: 1
        )
        XCTAssertEqual(output.count, 32)
    }

    func testScryptOutputLength() {
        let pass = Data("password".utf8)
        let salt = Data("salt".utf8)

        for length in [16, 24, 32, 64] {
            XCTAssertEqual(
                scrypt(password: pass, salt: salt, outputLength: length).count,
                length
            )
        }
    }
}

File: csharp/BCCrypto/BCCrypto.Tests/ScryptTests.cs
using BlockchainCommons.BCCrypto;

namespace BlockchainCommons.BCCrypto.Tests;

public class ScryptTests
{
    [Fact]
    public void TestScryptBasic()
    {
        byte[] pass = "password"u8.ToArray();
        byte[] salt = "salt"u8.ToArray();
        byte[] output = ScryptKdf.Derive(pass, salt, 32);
        Assert.Equal(32, output.Length);
        // Scrypt should be deterministic for same input
        byte[] output2 = ScryptKdf.Derive(pass, salt, 32);
        Assert.Equal(output, output2);
    }

    [Fact]
    public void TestScryptDifferentSalt()
    {
        byte[] pass = "password"u8.ToArray();
        byte[] salt1 = "salt1"u8.ToArray();
        byte[] salt2 = "salt2"u8.ToArray();
        byte[] out1 = ScryptKdf.Derive(pass, salt1, 32);
        byte[] out2 = ScryptKdf.Derive(pass, salt2, 32);
        Assert.NotEqual(out1, out2);
    }

    [Fact]
    public void TestScryptOptBasic()
    {
        byte[] pass = "password"u8.ToArray();
        byte[] salt = "salt"u8.ToArray();
        byte[] output = ScryptKdf.DeriveOpt(pass, salt, 32, 15, 8, 1);
        Assert.Equal(32, output.Length);
    }

    [Fact]
    public void TestScryptOutputLength()
    {
        byte[] pass = "password"u8.ToArray();
        byte[] salt = "salt"u8.ToArray();
        foreach (int len in new[] { 16, 24, 32, 64 })
        {
            byte[] output = ScryptKdf.Derive(pass, salt, len);
            Assert.Equal(len, output.Length);
        }
    }
}
